// @sanity-codegen/core/dist/index.standalone.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ResolveExpressionError", {
  enumerable: true,
  get: function () {
    return _.ResolveExpressionError;
  }
});
Object.defineProperty(exports, "SchemaParseError", {
  enumerable: true,
  get: function () {
    return _.SchemaParseError;
  }
});
Object.defineProperty(exports, "defaultStructures", {
  enumerable: true,
  get: function () {
    return _.defaultStructures;
  }
});
Object.defineProperty(exports, "generateQueryTypes", {
  enumerable: true,
  get: function () {
    return _.generateQueryTypes;
  }
});
Object.defineProperty(exports, "generateSchemaTypes", {
  enumerable: true,
  get: function () {
    return _.generateSchemaTypes;
  }
});
Object.defineProperty(exports, "schemaNormalizer", {
  enumerable: true,
  get: function () {
    return _.schemaNormalizer;
  }
});
Object.defineProperty(exports, "transformGroqToStructure", {
  enumerable: true,
  get: function () {
    return _.transformGroqToStructure;
  }
});
Object.defineProperty(exports, "transformSchemaToStructure", {
  enumerable: true,
  get: function () {
    return _.transformSchemaToStructure;
  }
});
Object.defineProperty(exports, "transformStructureToTs", {
  enumerable: true,
  get: function () {
    return _.transformStructureToTs;
  }
});
var _ = require("./");
