// @sanity-codegen/core/dist/default-structures/image.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.imagePaletteSwatchStructure = exports.imagePaletteStructure = exports.imageMetadataStructure = exports.imageHotspotStructure = exports.imageDimensionsStructure = exports.imageCropStructure = exports.imageAssetStructure = void 0;
var _utils = require("../utils");
const imageDimensionsStructure = (0, _utils.createStructure)({
  type: 'Object',
  canBeNull: false,
  canBeOptional: false,
  properties: [{
    key: '_type',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: 'sanity.imageDimensions'
    })
  }, {
    key: 'aspectRatio',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'height',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'width',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }]
});
exports.imageDimensionsStructure = imageDimensionsStructure;
const imagePaletteSwatchStructure = (0, _utils.createStructure)({
  type: 'Object',
  canBeNull: false,
  canBeOptional: false,
  properties: [{
    key: '_type',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: 'sanity.imagePaletteSwatch'
    })
  }, {
    key: 'background',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'foreground',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'population',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'title',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }]
});
exports.imagePaletteSwatchStructure = imagePaletteSwatchStructure;
const imagePaletteStructure = (0, _utils.createStructure)({
  type: 'Object',
  canBeNull: false,
  canBeOptional: false,
  properties: [{
    key: '_type',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: 'sanity.imagePalette'
    })
  }, {
    key: 'darkMuted',
    value: imagePaletteSwatchStructure
  }, {
    key: 'darkVibrant',
    value: imagePaletteSwatchStructure
  }, {
    key: 'dominant',
    value: imagePaletteSwatchStructure
  }, {
    key: 'lightMuted',
    value: imagePaletteSwatchStructure
  }, {
    key: 'lightVibrant',
    value: imagePaletteSwatchStructure
  }, {
    key: 'muted',
    value: imagePaletteSwatchStructure
  }, {
    key: 'vibrant',
    value: imagePaletteSwatchStructure
  }]
});
exports.imagePaletteStructure = imagePaletteStructure;
const imageMetadataStructure = (0, _utils.createStructure)({
  type: 'Object',
  canBeNull: false,
  canBeOptional: false,
  properties: [{
    key: '_type',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: 'sanity.imageMetadata'
    })
  }, {
    key: 'dimensions',
    value: imageDimensionsStructure
  }, {
    key: 'hasAlpha',
    value: (0, _utils.createStructure)({
      type: 'Boolean',
      canBeNull: false,
      canBeOptional: false
    })
  }, {
    key: 'isOpaque',
    value: (0, _utils.createStructure)({
      type: 'Boolean',
      canBeNull: false,
      canBeOptional: false
    })
  }, {
    key: 'lqip',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'palette',
    value: imagePaletteStructure
  }]
});
exports.imageMetadataStructure = imageMetadataStructure;
const imageAssetStructure = (0, _utils.createStructure)({
  type: 'Object',
  canBeNull: false,
  canBeOptional: false,
  properties: [{
    key: '_type',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: 'sanity.imageAsset'
    })
  }, {
    key: 'assetId',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'extension',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'metadata',
    value: imageMetadataStructure
  }, {
    key: 'mimeType',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'originalFilename',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'path',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'sha1hash',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'size',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'uploadId',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'url',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }]
});
exports.imageAssetStructure = imageAssetStructure;
const imageCropStructure = (0, _utils.createStructure)({
  type: 'Object',
  canBeNull: false,
  canBeOptional: true,
  properties: [{
    key: '_type',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: 'sanity.imageCrop'
    })
  }, {
    key: 'bottom',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'left',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'right',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'top',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }]
});
exports.imageCropStructure = imageCropStructure;
const imageHotspotStructure = (0, _utils.createStructure)({
  type: 'Object',
  canBeNull: false,
  canBeOptional: true,
  properties: [{
    key: '_type',
    value: (0, _utils.createStructure)({
      type: 'String',
      canBeNull: false,
      canBeOptional: false,
      value: 'sanity.imageHotspot'
    })
  }, {
    key: 'height',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'width',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'x',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }, {
    key: 'y',
    value: (0, _utils.createStructure)({
      type: 'Number',
      canBeNull: false,
      canBeOptional: false,
      value: null
    })
  }]
});
exports.imageHotspotStructure = imageHotspotStructure;
