// @sanity-codegen/core/dist/default-structures/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultStructures = void 0;
Object.defineProperty(exports, "imageAssetStructure", {
  enumerable: true,
  get: function () {
    return _image.imageAssetStructure;
  }
});
Object.defineProperty(exports, "imageCropStructure", {
  enumerable: true,
  get: function () {
    return _image.imageCropStructure;
  }
});
Object.defineProperty(exports, "imageDimensionsStructure", {
  enumerable: true,
  get: function () {
    return _image.imageDimensionsStructure;
  }
});
Object.defineProperty(exports, "imageHotspotStructure", {
  enumerable: true,
  get: function () {
    return _image.imageHotspotStructure;
  }
});
Object.defineProperty(exports, "imageMetadataStructure", {
  enumerable: true,
  get: function () {
    return _image.imageMetadataStructure;
  }
});
Object.defineProperty(exports, "imagePaletteStructure", {
  enumerable: true,
  get: function () {
    return _image.imagePaletteStructure;
  }
});
Object.defineProperty(exports, "imagePaletteSwatchStructure", {
  enumerable: true,
  get: function () {
    return _image.imagePaletteSwatchStructure;
  }
});
var _image = require("./image");
const defaultStructures = [_image.imageAssetStructure, _image.imageDimensionsStructure, _image.imageMetadataStructure, _image.imagePaletteStructure, _image.imagePaletteSwatchStructure, _image.imageCropStructure, _image.imageHotspotStructure];
exports.defaultStructures = defaultStructures;
