// @sanity-codegen/core/dist/utils/hash.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.objectHash = objectHash;
exports.unorderedHash = unorderedHash;
var _objectHash = _interopRequireDefault(require("object-hash"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function replacer(n) {
  // don't consider the contents of the function
  if (typeof n === 'function') {
    // this i just a random string that won't collide with anything
    return 'aaad5ec23bd8c7b1268fb02791c9a19ae1a43abb';
  }
  return n;
}
const objectHashCache = new WeakMap();

/**
 * a very simple object hash function powered by `object-hash`
 */
function objectHash(n) {
  const key = typeof n === 'object' && n ? n : undefined;
  if (key) {
    const cached = objectHashCache.get(key);
    if (cached) return cached;
  }
  const result = (0, _objectHash.default)({
    sanityCodegen: n
  }, {
    algorithm: 'md5',
    replacer,
    encoding: 'base64'
  }).replace(/\W/g, '').substring(0, 16).padStart(16, '0');
  if (key) {
    objectHashCache.set(key, result);
  }
  return result;
}

/**
 * takes any number of items and returns a hash where the top-level order is not
 * considered
 */
function unorderedHash(items) {
  return objectHash({
    sanityCodegenUnordered: items.map(objectHash).sort((a, b) => a.localeCompare(b, 'en'))
  });
}
